GH-34278: [C++] Expose schema in named table provider

icexelloss opened the pull request:
<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

In the case of PARQUET issues on JIRA the title also supports:

    PARQUET-${JIRA_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->
https://github.com/apache/arrow/issues/34278
### What changes are included in this PR?
This PR adds an parameter "schema" in the `NamedTableProvider` interface

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

### Are these changes tested?
Added test to check schema in named table provider
<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

### Are there any user-facing changes?
This is a developer API change.
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
<!-- **This PR includes breaking changes to public APIs.** -->

<!--
Please uncomment the line below (and provide explanation) if the changes fix either (a) a security vulnerability, (b) a bug that caused incorrect or invalid data to be produced, or (c) a bug that causes a crash (even when the API contract is upheld). We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **This PR contains a "Critical Fix".** -->

icexelloss commented:
cc @westonpace This should be ready for review

icexelloss commented on cpp/src/arrow/engine/substrait/serde_test.cc:
```diff
@@ -3200,21 +3207,22 @@ TEST(SubstraitRoundTrip, JoinRelWithEmit) {
-            std::make_shared<compute::TableSourceNodeOptions>(std::move(output_table));
-        return compute::Declaration("table_source", {}, options, "mock_source");
-      };
+  NamedTableProvider table_provider = [left_table, right_table](
```
Only interface and format change here

icexelloss commented on cpp/src/arrow/engine/substrait/serde_test.cc:
```diff
@@ -4114,21 +4123,21 @@ TEST(Substrait, SetRelationBasic) {
-    return compute::Declaration("table_source", {}, options, "mock_source");
-  };
+  NamedTableProvider table_provider =
+      [table1, table2](const std::vector<std::string>& names, std::shared_ptr<Schema>) {
```
Only interface and format change here

icexelloss commented on cpp/src/arrow/engine/substrait/serde_test.cc:
```diff
@@ -3049,21 +3055,22 @@ TEST(SubstraitRoundTrip, JoinRel) {
-            std::make_shared<compute::TableSourceNodeOptions>(std::move(output_table));
-        return compute::Declaration("table_source", {}, options, "mock_source");
-      };
+  NamedTableProvider table_provider = [left_table, right_table](
```
Only interface and format change here

icexelloss commented on cpp/src/arrow/engine/substrait/serde_test.cc:
```diff
@@ -2309,11 +2310,16 @@ TEST(SubstraitRoundTrip, FilterNamedTable) {
     if (table_names != names) {
       return Status::Invalid("Table name mismatch");
     }
+    if (!dummy_schema->Equals(*schema)) {
```
Added checks here to ensure schema is passed in correctly.

westonpace commented on cpp/src/arrow/engine/substrait/options.h:
```diff
@@ -65,8 +65,8 @@ enum class ARROW_ENGINE_EXPORT ConversionStrictness {
-using NamedTableProvider =
-    std::function<Result<compute::Declaration>(const std::vector<std::string>&)>;
+using NamedTableProvider = std::function<Result<compute::Declaration>(
+    const std::vector<std::string>&, const std::shared_ptr<Schema>)>;
```
Please use `const Schema&`.  `shared_ptr` should only be needed if the caller is likely to extend ownership.  In the rare case that is needed the caller can just make their own copy.

If you do think it is likely that caller will need to extend ownership you can use `const std::shared_ptr<Schema>&` (`const &` and not just `const`).

icexelloss replied:
Thanks sense. Will push another revision.

icexelloss replied:
Changed to `const Schema&`

westonpace reviewed (changes requested):
Looks good. Just one small change.

westonpace commented on python/pyarrow/_substrait.pyx:
```diff
@@ -27,7 +27,7 @@ from pyarrow.includes.libarrow cimport *
 
 
-cdef CDeclaration _create_named_table_provider(dict named_args, const std_vector[c_string]& names):
+cdef CDeclaration _create_named_table_provider(dict named_args, const std_vector[c_string]& names, const CSchema& schema):
```
I suppose we would eventually want to pass this schema on to the python function correct?

Something like:

```
    c_schema_sp = shared_ptr[CSchema](schema)
    py_schema = wrap_schema(c_schema_sp)

    py_table = named_args["provider"](py_names, py_schema)
```

Do you not need this for your use case?

icexelloss replied:
Yeah the python side support for this is left out but we only use named table provider via the C++ API.

I didn't do the python side because (1) we don't use it and (2) I am not very familiar with the python C extension. (Took me a while to figure the current C extension changes just to get it working).

icexelloss replied:
I can spend a bit time to see if I can get this working on the Python side. Looks like your example code is simple enough.

icexelloss replied:
Hmm, any quick ideas on the error?

```
    py_names = []
    for i in range(names.size()):
        c_name = names[i]
        py_names.append(frombytes(c_name))
    c_schema_sp = shared_ptr[CSchema](schema)
                                    ^
------------------------------------------------------------

pyarrow/_substrait.pyx:42:37: Cannot convert 'shared_ptr[CSchema]' to Python object
gmake[2]: *** [CMakeFiles/_substrait_pyx.dir/build.make:71: CMakeFiles/_substrait_pyx] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:785: CMakeFiles/_substrait_pyx.dir/all] Error 2
```

icexelloss replied:
Hmm I might be importing `wrap_schema` from the wrong place:

```
from pyarrow.lib import frombytes, tobytes, wrap_schema
```

I think it's complaining about 
```
py_schema = wrap_schema(c_schema_sp)
```
(error trace is weird on the wrong line)

icexelloss replied:
Ok I managed to get extesion to compile with

```
c_in_schema = make_shared[CSchema](move(schema))
```

Let me fix  the python test next

westonpace replied:
Hmm, the first error was probably that I didn't point out that you would need to add `c_schema_sp` to your `cdef` section (so cython was recognizing `c_schema_sp` as a python object and not a C++ object).

icexelloss replied:
Actually - hold off - I think I didn't handle ownership correctly - should be probably doing a copy instead of move here. Let me see if I can fix.

icexelloss replied:
Ok I pushed another revision - I think this is correct now

icexelloss commented:
@westonpace pushed another revision
* Fixed lint
* Addressed https://github.com/apache/arrow/pull/34279#discussion_r1114657183 and exposed schema via python named table provider (also added test and updated the doc)

westonpace reviewed (approved):
Thanks for covering the bindings as well.
